Rightscale development v1

lopaka opened the pull request.

Commit:
add kitchen.yml file

Commit:
Added Berksfile

Commit:
Merge pull request #1 from lopakadelp/white_14_02_acu151058_dns_cookbook

White 14 02 acu151058 dns cookbook

Commit:
Update kitchen.yml attributes. Update providers/default.rb to pass hash with symbol keys to be read correctly by fog.  Update version of Fog.

Commit:
update .kitchen.yml file to use dnsmadeeasy

Commit:
added dnsmadeeasy provider

Commit:
create update recipe

Commit:
add record_id and name/fqdn to test in kitchen.yml.  Added update action to default provided, not yet available.

Commit:
Added update test in .kitchen.yml

Commit:
Provider is already an attribute given to Fog, so easier to do logic in default provider in dns cookbook.

Commit:
Update code to handle DNSMadeEasy's lack of an update record method at the moment

Commit:
Additional code to handle DNSMadeEasy's lack of an update record method

Commit:
Support for CentOS using correct names of package dependencies for nokogiri

Commit:
Use correct family name, debian instead of ubuntu.

Commit:
Merge development to match what was used from forked branch.

Commit:
Updated create and destroy actions to only use specific cloud actions when needed

Commit:
Create and destroy actions to handle multiple entires with same name entry.

Commit:
Improvements on update action.

Commit:
Delete and create entry on updating a DNSMadeEasy record

Commit:
Set update recipe with oldvalue attribute

Commit:
Set ttl in attributes to 1800 seconds (30 minutes)

Commit:
Improve comments.

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
+  # Will check zone.id and zone.domain based on known providers:
+  # DME - zone.id matches domain
+  # AWS - zone.domain matches domain
+  unless zone = connection.zones.detect { | zone_entry |
```
No space around `|`. I don't see any different coding style in the remaining code.

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
+  if zone.records.detect { | record_entry |
+    (record_entry.name == subdomain || record_entry.name =~ /^#{subdomain}\.#{domain}\.{0,1}$/) &&
+    record_entry.value == new_resource.entry_value
+  }
```
Please use `do..end` for multiline blocks.

lopaka replied:
This is actually a single line broken by the 120char rule.  Should we break the 120char rule since it's not often I see a `do..end` block for a single command broken to multiple lines, especially in enumerable methods?

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
-    new_resource.updated_by_last_action(true)
   end
+
+  new_resource.updated_by_last_action(true)
```
This line should go inside the `else` block (where the record doesn't exist) or else it will always notify other subscribed resources.

arangamani commented on resources/default.rb:
```diff
@@ -1,8 +1,9 @@
 
 attribute :entry_name, :kind_of => String
 attribute :entry_value, :kind_of => String, :required => true
+attribute :entry_oldvalue, :kind_of => String, :default => ''
```
Is this default value necessary? No other attributes in this resource have default value.

lopaka replied:
This was necessary during dev to handle the `unless new_resource.entry_oldvalue.empty?` condition.  But best to fix it up in the provider.

Also, there is a default for `:type`.

arangamani replied:
Related to [this](https://github.com/lopakadelp/dns/pull/3#discussion_r9913176).

arangamani commented on recipes/update.rb:
```diff
@@ -0,0 +1,12 @@
+  action :update
+end
+
+
```
We don't need these 2 extra lines.

arangamani commented on recipes/fog.rb:
```diff
@@ -2,14 +2,17 @@
+  ["centos", "rhel"] => ["libxslt-devel", "libxml2-devel"],
+  "debian" => ["libxslt-dev", "libxml2-dev"]
+  )
+nokogiri_dependencies.each do |pkg|
```
Just a thought. Would it make sense to use the community [xml cookbook](http://community.opscode.com/cookbooks/xml) here? It has support for more platforms and provides a way to install packages in compile time.

``` ruby
node.override['xml']['compiletime'] = true
include_recipe 'xml'
```

lopaka replied:
I will need to look into this.  I just added the `value_for_platform_family` section to do minimum changes the cookbook to support CentOS.

lopaka replied:
Yep...looks like this may work.  Going to add depend on `xml`.

arangamani commented on recipes/delete.rb:
```diff
@@ -0,0 +1,7 @@
+include_recipe 'dns::fog'
+
+dns 'Delete DNS entry' do
+  entry_name lazy{ node[:dns][:entry][:name] }
```
A space between `lazy` and the block.

lopaka replied:
This was an altered copy from the `default` recipes in the the original cookbook, so was just following copying what was already done.  Should we update that as well?

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
+    Chef::Log.info "Deleted DNS entry(s) #{new_resource.entry_name}"
   end
+
+  new_resource.updated_by_last_action(true)
```
This line should go inside the `else` block as well.

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
+    zone_entry.id =~ /^#{domain}\.{0,1}$/ || zone_entry.domain =~ /^#{domain}\.{0,1}$/
+  }
+    raise "DOMAIN '#{domain}' NOT AVAILABLE IN ACCOUNT"
+  end
```
This verification is done in all actions so may be make this a reusable method?

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
+  end
+
+  Chef::Log.info "Updated DNS entry #{new_resource.entry_name}"
+  new_resource.updated_by_last_action(true)
```
Is this action idempotent? Is it possible to check the current value and compare it with the new value?

lopaka replied:
From reading http://docs.opscode.com/lwrp_custom_provider.html#updated-by-last-action, this allows notification to be triggered from LWRP actions.  So, this should not care about the last updated action.

arangamani replied:
I'm sorry if I wasn't clear. This action will always notify other resources if a matching record is found regardless of its current and new values. What I meant was to check the current values of the record with the new values provided and then update the record only when it is necessary. With current implementation, if I use this action in my recipe and notify other resources (for example, restart of mysql service), it will always be notified even if the value is not changed.

lopaka replied:
I understand now.  For the 'update' method, it will raise if the IP is being changed back to the same.  However, I believe we should just exit quietly if the update is using the current settings.

arangamani replied:
Correct, it shouldn't raise. It should simply log and exit.

arangamani commented on providers/default.rb:
```diff
@@ -5,53 +5,171 @@ def load_current_resource
+  }
+
+  # Narrow search if existing value is given.
+  unless new_resource.entry_oldvalue.empty?
```
Shouldn't we also check for `nil` values?

Commit:
A few coding style changes.  Moved setting of new_resource.updated_by_last_action(true) within conditions where updates are actually done.

Commit:
Follow original code and use do..end in enumerator blocks

Commit:
Created method to verify domain/zone.
Handles entry_oldvalue if nil.
Uses community 'xml' cookbook to install required dependencies for nokogiri.

Commit:
Simple coding standard fix.

Commit:
Change reserved name attribute :provider to :dns_provider

Commit:
Use dns_provider attribute in default, delete, and update recipes.

Commit:
Changed entry_oldvalue attribute to entry_currentvalue and use it in delete andupdate action.

Commit:
In update action, if new value equals current value, will not update anything.

Commit:
Update README.md and clean trailing white spaces

douglaswth commented on README.md:
````diff
@@ -1,26 +1,30 @@
 
 ```ruby
 dns 'dns.example.org' do
+  dns_provider 'dns_provider'
````
Maybe this should match what the previous value was somewhat. Something like `'some_dns_provider'`?

lopaka replied:
Agreed

douglaswth commented on .kitchen.yml:
```diff
@@ -0,0 +1,90 @@
+    box: opscode-centos-6.4
+    box_url: https://opscode-vm-bento.s3.amazonaws.com/vagrant/opscode_centos-6.4_provisionerless.box
+
+suites:
```
I don't see why we would contribute kitchen tests that don't actually work.

lopaka replied:
We shouldn't.  Currently working on the 'fake' cookbook and setting .kitchen.yml to work with Fog moch mode - which does not work with dnsmadeeasy provider.

douglaswth replied:
It would really make more sense to use chefspec rather than test-kitchen for doing testing that involves mocking.

lopaka replied:
Mocking currently works with test-kitchen.  We can add chefspec at a later time.

douglaswth replied:
I think we should set the credential values such that they are more obviously fake for mocking.

lopaka replied:
agreed.

douglaswth commented on README.md:
```diff
@@ -1,26 +1,30 @@
+  dns_provider 'dns_provider'
   credentials :some_cloud_token => '[TOKEN]', :some_cloud_key => '[KEY]'
-  provider 'some_cloud'
+  entry_name 'www.example.org'
```
Why are we adding this to the example? It already has a name of `'dns.example.org'`.

lopaka replied:
More dev residue - removing it.

douglaswth commented on README.md:
```diff
@@ -33,9 +37,10 @@ end
 * IRC: Freenode @ #heavywater
-* Cookbook: http://ckbk.it/dns
\ No newline at end of file
+* Cookbook: http://ckbk.it/dns
```
What changed on this line?

lopaka replied:
```
> * Cookbook: http://ckbk.it/dns
\ No newline at end of file
```

douglaswth commented on Strainerfile:
```diff
@@ -0,0 +1,3 @@
+knife: bundle exec knife cookbook test $COOKBOOK
+foodcritic: bundle exec foodcritic -f any $SANDBOX/$COOKBOOK
+kitchen: bundle exec kitchen test --parallel
```
Does Heavywater already use Strainer? I don't think we should be polluting their stuff with it, especially since we're planning on moving away from it as well.

douglaswth commented on Berksfile:
```diff
@@ -0,0 +1,6 @@
+metadata
+
+cookbook "build-essential"
+cookbook "xml"
```
Both `build-essential` and `xml` are already specified in `metadata.rb`.

Commit:
Initial setup for mock testing

Commit:
Finish dev of kitchen testing and update provider from test results.

Commit:
Cleaned up Berksfile and removed unused Strainerfile

Commit:
Removed unneeded test-kitchen suite.  Gave test creditials a value to make it obvious that it is being used for mocking and testing.

nitinmohan87 commented on providers/default.rb:
```diff
@@ -1,57 +1,195 @@
+
+  else
+    # Options for record creation.
+    options = {:ttl => new_resource.ttl ? new_resource.ttl : 1800}
```
Why is `1800` hard coded here? Can we set the default value for `ttl` attribute in the resource itself?

lopaka replied:
Examples on DNSMadeEasy 2.0 and 1.2 API docs showed 1800 seconds.  But good idea on adding it to the resource.

nitinmohan87 commented on providers/default.rb:
```diff
@@ -1,57 +1,195 @@
+      zone.records.create({
+        :value => new_resource.entry_value,
+        :name => new_resource.entry_name,
+        :type => new_resource.type.upcase}.merge(options)
```
`}.merge(options)` should be in the next line. All occurrences of this should be changed.

nitinmohan87 commented on recipes/fog.rb:
```diff
@@ -1,15 +1,18 @@
 end
+
+# Turn on Fog mocking
+if node[:dns][:mock]
```
Can we do this in the `fake` cookbook and add the recipe to the Vagrant and Kitchen runlists?

Commit:
- Created fake fog_mock recipe to remove mock test setup from production cookbooks.
- Set TTL default in resource instead of inside provider logic.
- Small cosmetic coding standard changes.

nitinmohan87 commented on attributes/default.rb:
```diff
@@ -5,4 +5,9 @@
 default[:dns][:chef_client_config] = false
+
+# Used to enable mocking in Fog
+default[:dns][:mock] = false
```
I think you can move this attribute to the `fake` cookbook as well. Set this to `false` by default. This means you can also remove the input setting from the `.kitchen.yml`

nitinmohan87 replied:
Actually we won't need this attribute at all since there is a separate recipe to do the fog mocking.

Commit:
Remove unnecessary mocking node attribute.

nitinmohan87 commented on .kitchen.yml:
```diff
@@ -0,0 +1,43 @@
+    cloud:
+      provider: vagrant
+    dns:
+      mock: true
```
You can remove this as well.
